web/profiles/custom/os2loop/modules/os2loop_subscriptions/src/EventSubscriber/SubscriptionFlagSubscriber.php

```diff
@@ -0,0 +1,180 @@
+      // The message red flag.
+      $flag = $this->flagService->getFlagById('message_read');
+      // A list of messages currently displayed to the user @ /user/messages.
+      $active_messages = $this->getActiveMessages();
+
+      if ('os2loop_subscription_node' === $flag_id) {
+        $this->flagMessageRead($entity_id, $active_messages, $flag);
+      }
+
+      if ('os2loop_subscription_term' === $flag_id) {
+        $nodes = $this->getReferencedNodes($entity_id);
+        foreach ($nodes as $nid => $node) {
+          $this->flagMessageRead($nid, $active_messages, $flag);
+        }
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    $events = [];
+    $events[FlagEvents::ENTITY_FLAGGED][] = ['onFlag'];
+    return $events;
+  }
+
+  /**
+   * Get all messages that reference a specific node.
+   *
+   * @return array
+   *   A list of messages related to node.
+   */
+  private function getReferencedMessages(string $nid): array {
+    return $this->entityTypeManager
+      ->getListBuilder('message')
+      ->getStorage()
+      ->loadByProperties([
+        'os2loop_message_node_refer' => $nid,
+      ]);
+  }
+
+  /**
+   * Get all nodes that reference a specific term from subject or profession.
+   *
+   * @return array
+   *   A list of nodes related to a term.
+   */
+  private function getReferencedNodes(string $tid): array {
+    $subject_reference = $this->entityTypeManager
+      ->getListBuilder('node')
+      ->getStorage()
+      ->loadByProperties([
+        'os2loop_shared_subject' => $tid,
+      ]);
+
+    $profession_reference = $this->entityTypeManager
+      ->getListBuilder('node')
+      ->getStorage()
+      ->loadByProperties([
+        'os2loop_shared_profession' => $tid,
+      ]);
+
+    return $subject_reference + $profession_reference;
+  }
+
+  /**
+   * Get all active messages.
+   *
+   * Get all messages that the current user has not read from the
+   * os2loop_messages view.
+   *
+   * @return array
+   *   A list of message ids of messages that the user has not yet read.
+   */
+  private function getActiveMessages(): array {
+    $message_ids = [];
+    $view = Views::getView('os2loop_messages');
+    $view->execute();
+    foreach ($view->result as $row) {
+      $message_ids[] = $row->_entity->id();
+    }
+    return $message_ids;
+  }
+
+  /**
+   * Flag a message as being read.
+   *
+   * @param int $entity_id
+   *   A node id.
+   * @param array $active_messages
+   *   List of messages currently displayed to the user.
+   * @param \Drupal\flag\FlagInterface|null $flag
+   *   The flag we want to set.
+   */
+  private function flagMessageRead(int $entity_id, array $active_messages, FlagInterface $flag = NULL) {
+    $messages = $this->getReferencedMessages($entity_id);
+    foreach ($messages as $id => $message) {
+      // Check if the flag has already been set.
+      if (0 < $this->currentUser->id()) {
```

Comment:
Use `$this->currentUser->isAuthenticated()` here to make it more clear that we're only doing this for authenticated users.